### ISSUE TITLE
Kotlin+Dagger best practices/documentation

### ISSUE THREAD

**Issue by ronshapiro** [google/dagger#900]
Opening this as a tracking bug for all kotlin related documentation that we should be add/best practices that we should call out to make using Dagger w/ Kotlin easier.

One example: How to achieve the effect of `static @Provides` in Kotlin.

Feel free to comment new ideas, but don't make "me too" or "i agree with XYZ" comments.

**Comment by bejibx** [comment]
> How to achieve the effect of ```static @Provides``` in Kotlin.

Still writing my modules in java because of this.

**Comment by hzsweers** [comment]
If you have injected properties (as "fields"), qualifiers _must_ have `field:` designation.

Good

```kotlin
@Inject @field:ForApi lateinit var gson: Gson
```

Bad

```kotlin
@Inject @ForApi lateinit var gson: Gson // @ForApi is ignored!
```

**Comment by hzsweers** [comment]
Static provides can be achieved via `@JvmStatic`. There are two scenarios I see this come up:

top-level `object`s

```kotlin
@Module
object DataModule {
  @JvmStatic @Provides fun provideDiskCache() = DiskCache()
}
```

If you have an existing class module, things get a bit weirder

```kotlin
@Module
abstract class DataModule {
  @Binds abstract fun provideCache(diskCache: DiskCache): Cache

  @Module
  companion object {
    @JvmStatic @Provides fun provideDiskCache() = DiskCache()
  }
}
```

The way this works is as follows:
- the `companion object` must also be annotated as `@Module`
- under the hood, the kotlin compiler will duplicate those static provides methods into the `DataModule` class. Dagger will see those and treat them like regular static fields. Dagger will also see them in the companion object, but that "module" will get code gen from dagger but be marked as "unused". The IDE will mark this as such, as the `provideDiskCache` method will be marked as unused. You can tell IntelliJ to ignore this for annotations annotated with `@Provides` via quickfix

I sent Ron a braindump once of how dagger could better leverage this, i.e. no requirement for JvmStatic and just call through to the generated `Companion` class, but I think that's out of the scope of this issue :). I've been meaning to write up a more concrete proposal for how that would work.

**Comment by hzsweers** [comment]
Another gotcha is inline method bodies. Dagger relies heavily on return types to connect the dots. In kotlin, specifying the return type is optional sometimes when you can inline the method body. This can lead to confusing behavior if you're counting on implicit types, especially since the IDE will often try to coerce you into quickfixing them away

That is, you could take write in one of two ways

```kotlin
@Provides fun provideDiskCache() = DiskCache()

@Provides fun provideDiskCache(): Cache {
  return DiskCache()
}
```

The first function is valid, but `DiskCache` is what's on the DI graph there because that's the inferred return type.

The second function is also valid, but now `Cache` is what's on the graph and `DiskCache` is just an implementation detail.

The IDE will try to suggest inlining the second one. You can do so, but be mindful of potentially changing return types.

**Comment by JakeWharton** [comment]
An expression body doesn't preclude the use of a return type.

**Comment by hzsweers** [comment]
Good point, tweaked the wording to make it clear that it's only if you drop the return type

**Comment by tasomaniac** [comment]
`@JvmSuppressWildcards` is incredibly useful when you are injecting classes with generics. Can be handy when using multimap injection.

**Comment by jhowens89** [comment]
What a great thread! I was fighting this fight this last weekend. Here is the syntax for injection that's both qualified and nullable:

`@field:[Inject ChildProvider] @JvmField var coordinatorProvider: CoordinatorProvider? = null`

**Comment by heinrichreimer** [comment]
@tasomaniac #807 has also cost me quite some time to debug.
Doesn't look nice but at least it works:
```kotlin
@Inject lateinit var foo: Set<@JvmSuppressWildcards Foo>
```
_(taken from https://stackoverflow.com/a/43149382/2037482)_

**Comment by drawers** [comment]
This may seem obvious for experienced users but Kotlin beginners with Dagger 2 could trip up on the syntax for constructor injection that requires the explicit use of the otherwise uncommon `constructor` keyword:

e.g.,
`class Foo @Inject constructor (private val bar: Bar)
`

**Comment by charlesdurham** [comment]
As an  add-on from @heinrichreimer, the same thing is required when injecting functions that have parameters:

```
... @Inject constructor(val functionalThing: @JvmSuppressWildcards(true) (String) -> String)
```

**Comment by InvisibleGit** [comment]
With Kotlin 1.2 we can use array literals in annotations, ditching `arrayOf()` call in them

`@Component(modules = [LibraryModule::class, ServicesModule::class])`

**Comment by guelo** [comment]
As far as I can tell Dagger is unable to inject Kotlin functions. This fails

```kotlin
@Module
class Module() {
  @Provides fun adder(): (Int, Int) -> Int = {x, y -> x + y}
}

class SomeClass {
  @Inject lateinit var adder:  (Int, Int) -> Int

  fun init() {
    component.inject(this)
    println("${adder(1, 2)} = 3")
  }
}
```
`error: kotlin.jvm.functions.Function2<? super java.lang.Integer,? super java.lang.Integer,java.lang.Integer> cannot be provided without an @Provides- or @Produces-annotated method.`

Theoretically you could provide a kotlin.jvm.functions.Function2 but I failed to make it work. Even with this
```kotlin
@Provides
fun adder(): kotlin.jvm.functions.Function2<Integer, Integer, Integer> {
  return object : kotlin.jvm.functions.Function2<Integer, Integer, Integer> {
    override fun invoke(p1: Integer, p2: Integer): Integer {
      return Integer(p1.toInt() + p2.toInt())
    }
  }
}
```
it still says a Function2 cannot be provided.

**Comment by JakeWharton** [comment]
You need to add @JvmSuppressWildcards at the injection site to ensure the
signature matches.

**Comment by gildor** [comment]
I understand that this issue about "best practices", but on Reddit AMA @ronshapiro mentioned that this issue the place to collect kotlin-specific features.
There is a couple possible kotlin-specific features that require work with Kotlin Metadata, but would be very useful for Kotlin projects and make dagger usage even more pleasant

1. Support `typealias` as an alternative for @Named and custom @Qualifier annotations. It's completely compiler feature, but you can get typealias name from @Metadata annotation, so can use it to detect different qualifiers of the same class
2. Support nullability as an alternative for Optional value. But I see some drawbacks: you cannot distinguish between non-initilized nullable and inilized one, but it's still can be a usefult and efficient replacement for Optional dependencies.
3. Native support of lambda injection that doesn't require to use `@JvmSuppressWildcards`. This proposal can be extended to other generics, like for Lists and so on, but it not always required behavior, but in the case of lambdas I don't see any good reason do not consider any lambda (kotlin.jvm.functions.* interface) as generic without a wildcard. It will make lambda injection much less wordy. Also, works pretty well with `typealias` as qualifier.

**Comment by ronshapiro** [comment]
It would probably be good to make sure that no binding method used default parameters. It wouldn't make sense, and it definitely would be misleading.

**Comment by tobeylin** [comment]
The method injection works well on Kotlin setter.
```
@set: Inject
lateinit var booleanProvider: Provider<Boolean>
```
If you need the qualifier, must write the custom setter.
```
var booleanProvider: Provider<Boolean>? = null
        @Inject set(@MyQualifier value){
            field = value
}
```

**Comment by sophiataskova** [comment]
@hzsweers re: your comment on static provides methods in Kotlin, does the Kotlin implementation offer the same performance wins as "true" Java static methods? Does it provide any performance wins? What's your main reason for making your modules this way?


> Static provides can be achieved via @JvmStatic. There are two scenarios I see this come up:
> 
> top-level objects
> 
> @Module
> object DataModule {
>   @JvmStatic @Provides fun provideDiskCache() = DiskCache()
> }
> If you have an existing class module, things get a bit weirder
> 
> @Module
> abstract class DataModule {
>   @Binds abstract fun provideCache(diskCache: DiskCache): Cache
> 
>   @Module
>   companion object {
>     @JvmStatic @Provides fun provideDiskCache() = DiskCache()
>   }
> }
> The way this works is as follows:
> 
> the companion object must also be annotated as @Module
> under the hood, the kotlin compiler will duplicate those static provides methods into the DataModule class. Dagger will see those and treat them like regular static fields. Dagger will also see them in the companion object, but that "module" will get code gen from dagger but be marked as "unused". The IDE will mark this as such, as the provideDiskCache method will be marked as unused. You can tell IntelliJ to ignore this for annotations annotated with @Provides via quickfix
> I sent Ron a braindump once of how dagger could better leverage this, i.e. no requirement for JvmStatic and just call through to the generated Companion class, but I think that's out of the scope of this issue :). I've been meaning to write up a more concrete proposal for how that would work.

**Comment by hzsweers** [comment]
static provides allow you to make your modules truly stateless and allows you to simplify plumbing (no instance wiring required). The biggest win there is basically architecture. The "performance" wins always seemed kind of a reach. 25% of something measured in nanoseconds on a non-hotpath seems a nice-to-have rather than a strong motivating factor.

**Comment by ronshapiro** [comment]
The `invokestatic` might not be the true benefit - not needing the module instance allows all static provides to get horizontally merged, eliminating the class loading of all of the module classes.

**Comment by arekolek** [comment]
@sophiataskova See the [discussion I had with Jeff Bowman on this topic](https://stackoverflow.com/a/47701934/1916449), maybe you'll find it useful.

**Comment by tasomaniac** [comment]
My 2 cents:

If somebody put a gun to your head to do 100% Kotlin, sure this is a solution. But I really prefer Java Modules to this. It is totally fine for me to 1 or 2 Java abstract classes in my feature. No inner classes, single flat file of provisions. Much simpler 🎉

**Comment by sophiataskova** [comment]
Should I take it that Kotlin does not offer any of the advantages of `static @Provides` that you get in Java? 

Ron says that `not needing the module instance allows all static provides to get horizontally merged, eliminating the class loading of all of the module classes` but both Kotlin implementations offered here seem to still involve class loading for each module. Again, I kinda hope I'm wrong, and can make my app faster without converting to Java.

**Comment by JakeWharton** [comment]
First of all, this speed benefit will be extraordinarily minor. If you are looking to make your app faster then use a profiler. You will find 100 easier optimization targets.

Beyond that, don't use `companion object` for modules. Use `object`. In that case the instance will be unused and its initialization code will be removed by R8 and the methods will be truly static and can also be inlined just like Java.

**Comment by tasomaniac** [comment]
It is unfortunately not possible to have abstract methods in object. You would need 2 modules. I try not to mix abstract and static provisions anyways. But if you need to, object is not a solution. This JvmStatic trick allows you to have them in 1 module.

**Comment by JakeWharton** [comment]
I think you should ignore that person!

The last point isn't tied to any language and should probably be considered
separately.

**Comment by tasomaniac** [comment]
It is easier to have only 1 constructor in Kotlin since they have default value support. Dagger does not understand that and that's why often times I find myself having secondary constructor with Inject annotation to provide defaults. 

If Dagger is going to get a Kotlin specific features, I think understanding the default values provided would be my favorite.

**Comment by cgruber** [comment]
The problem with @Inject on the type is that it isn't JSR-330. I like the
idea, but I think it breaks the language.  (Not that Dagger couldn't stray,
but it should be considered carefully).

An alternative would be to just accept single-constructors as implicitly
@Inject, which I think @warabei14 refers to above, but that walks back a
rather significant decision to not support implicit constructors (mostly to
avoid stuff like injecting a newly constructed empty String because it has
a default constructor).   There are subtleties here, and I like the
possible feature set for Kotlin, but we should make sure we represent those
earlier concerns properly in the discussion, so we don't trample older
design decisions that may still be meaningful.

**Comment by hzsweers** [comment]
You could have a separate artifact of kotlin-specific annotations that, when used, require Kotlin `@Metadata` annotations to be present too. In the case of `@Inject` on a class, kotlin metadata annotations can tell you what the `primary` constructor is of a given class, and you'd then just want to wire that up in Dagger as a fake `@Inject` constructor. That'd allow you to have something like `@KotlinInject` that doesn't tread on the existing JSR330 annotations.

**Comment by JakeWharton** [comment]
I'm still not convinced we need to invent both a problem and a solution. Is
`@Inject constructor` really that bad? Why would someone describe a feature
of the language as unidiomatic unless they were just trying to code golf
character count?

**Comment by cgruber** [comment]
I'm with Jake here.  I don't accept that it isn't idiomatic kotlin either.
If you need to see the constructor, this is how you do it.  I've done this
several times and it reads just fine to me.

**Comment by tasomaniac** [comment]
Here is how you would add Dagger's annotation processor options with Kotlin's `kapt`. It looks pretty nice actually.

```
kapt {
  arguments {
    arg('dagger.formatGeneratedSource', 'disabled')
  }
}
```

You can use ☝️ for all the options mentioned here: https://google.github.io/dagger/compiler-options.html

**Comment by ultraon** [comment]
I created interesting the config for kapt with different optimization options in this way:
```Groovy
kapt {
    useBuildCache = false
    // If you have some issues with kapt to get more meaningful message in terminal set to "true" "mapDiagnosticLocations" and "correctErrorTypes" below
    mapDiagnosticLocations = false
    correctErrorTypes = false

    javacOptions {
        // Dagger2 APT optimizations, see https://google.github.io/dagger/compiler-options
        option('-Adagger.fastInit=enabled')
        if (isCiBuild) {
            option('-Adagger.moduleBindingValidation=WARNING')
            // Increase the max count of errors from annotation processors, default is 100,
            // see https://kotlinlang.org/docs/reference/kapt.html#java-compiler-options
            option("-Xmaxerrs", 500)
        } else {
            option('-Adagger.formatGeneratedSource=disabled')
            option('-Adagger.gradle.incremental')
            option('-Adagger.moduleBindingValidation=ERROR')
        }
    }
}
```

**Comment by tasomaniac** [comment]
You should consider enabling the build cache with kapt to improve the build speed. This is a huge speed up especially in highly modularized apps. We have been using it with Dagger for quite some time with no problem. 

```
kapt {
    useBuildCache = true
}
```

**Comment by matejdro** [comment]
Does `option('-Adagger.gradle.incremental')` even make any difference considering kapt is not incremental?

**Comment by vRallev** [comment]
I found another little workaround. If you have an `internal` function for multi bindings, then Dagger can't find the method, because the Kotlin compiler mangles the name.
```kotlin
@Module
abstract class MyModule {
  // methods

  @Module
  companion object {
    @Provides @IntoSet @JvmStatic
    internal fun contributeInternalObject(object: InternalClass): InterfaceClass = object
  }
}
```
This will fail with `error: cannot find symbol`. 

The trick is to use `@PublishedApi`. That tells the Kotlin compiler to skip the name mangling and makes Dagger happy.
```kotlin
@Provides @IntoSet @JvmStatic @PublishedApi
internal fun contributeInternalObject(object: InternalClass): InterfaceClass = object
```

**Comment by prasanthmurugan** [comment]
Hi, i am learning dagger dependency injection. But got stuck with the below error for 2 days. I have create StackOverflow post regarding kindly check the below link.

Any help or suggestion could be great !! 

[StackOverflow post link](https://stackoverflow.com/q/55892748/4111151)

**Comment by ZacSweers** [comment]
@prasanthmurugan that doesn't have anything to do with Kotlin. Let's keep this issue on topic

@matejdro kapt supports incremental apt as of kotlin 1.3.30 (though theres a Gradle flag you need to enable)

**Comment by prasanthmurugan** [comment]
@ZacSweers thanks ! I am new to dagger can you help me out on why provider from Activitymodule is not getting injected in ViewModelModule in my case https://stackoverflow.com/q/55892748/4111151. Dagger is throwing an dependency exist error but i don't know how to match both.

**Comment by cliveleehere** [comment]
Hm when there's a class delegation, e.g.:
```
interface A {
  fun a()
}
```
and
```
class AImpl @Inject constructor(): A {
  fun a() { ... }
}
``` 
(Note that AImpl is bound to A somewhere)
```
class C @Inject ( injectedA : A) : A by injectedA 
```

Then, dagger throws an error during kapt.
However, if the class C is changed to _not_ use class delegation:
```
class C @Inject (injectedA : A) : A {
  fun a() { injectedA.a() }
}
```
everything works D:

**Comment by ZacSweers** [comment]
@ronshapiro @gildor Another thought on default parameters: this could allow for otherwise "missing" dependencies when composing modules. Consider this:

```kotlin
class HttpModule {
  @Provides
  fun provideHttpClient(cache: Cache? = Cache.inMemoryCache()): HttpClient {
    return HttpClient(cache)
  }
}

@Module(includes = [HttpModule::class]
class TwitterModule {
  @Provides
  fun provideTwitterClient(httpClient: HttpClient): TwitterClient {
    return TwitterClient(httpClient)
  }
}

//
// TwitterModule could be used with any of the below AppModule variants
//

// Cache is defined and not null, HttpClient will end up with RealCache
@Module(includes = [TwitterModule::class]
class ProductionAppModule {
  @Provides fun provideCache(): Cache {
    return RealCache()
  }
}

// No cache is defined, HttpClient will end up with default inMemoryCache()
@Module(includes = [TwitterModule::class]
class IntegrationTestAppModule {
  
}

// Cache is defined but null, HttpClient will end up with no cache
@Module(includes = [TwitterModule::class]
class ConstrainedEnvironmentAppModule {
  @Provides fun provideCache(): Cache? = null
}
```

You _can_ model absence and nullability in Kotlin at least. I'm not familiar enough with dagger to say off the top of my head if it differentiates between null and absent, but I'd assume it does with Optional. Then the above just boils down to invoking the defaults method and indicating if the `cache` parameter is present, rather than modeling with something like Optional. Invoking the synthetic defaults method would require reflection with today's build tooling, but could be achieved with bytecode generation once gradle and kapt properly support Filer-generated bytecode. The above approach is definitely advanced usage, but I think a really powerful affordance of the language. This can also be done with constructors for constructor injection.

**Comment by tasomaniac** [comment]
In a Dagger component, an instance is only ever available once. You don't have 2 different ways of injecting the same instance. If I'm not wrong, Kotlin default value support would introduce that. That means, it would not be clear which instance is injected in a given graph.

**Comment by gmale** [comment]
Should these best practices be summarized into a document, perhaps as part of the [initiative to improve Dagger on Android](https://www.youtube.com/watch?v=o-ins1nvbDg)? That might allow this issue to be closed, rather than remain open indefinitely.

**Comment by ZacSweers** [comment]
The purpose of this issue is to be a living doc, I don’t think moving it into a frozen documentation that the community can’t contribute to is good

**Comment by kvithayathil** [comment]
Definitely like the idea of keeping this issue open for all contributors. It would be great to transpose of these best practises somewhere indexable/searchable

**Comment by dmitry-ryadnenko-izettle** [comment]
Can we expect at some point to be able to add `@Inject` annotations to top-level factory functions like we can do with class constructors? Example.

```kotlin
@Inject
fun makeMyDependency(): MyDependency = object : MyDependency {

}
```

So, in Dagger, you can tell Dagger how to resolve a dependency by dropping an @Inject annotation on the class constructor.

```kotlin
class MyDependencyImpl @Inject constructor() : MyDependency {
    
}
```

What I'm missing really badly is a way to do the same thing but with top-level factory functions.

```kotlin
@Inject
fun makeMyDependency(): MyDependency = object : MyDependency {

}
```

One of the most important scenarios where I need this is to be able to tell Dagger how to provide a function.


```kotlin
@Inject
fun makeMyFunction(): () -> Something = { /* implement function */ }
```

Functions doesn't have constructors, so when using a factory function like the one above there is no way to benefit from boilerplate generation that we have with `@Inject` in class constructors. So when providing functions we always have to write boilerplate like this.

```kotlin
@Module
class MyFunctionModule {
    @Binds
    fun provideMyFunction(
            dependency1: Dependency1,
            ...
            dependencyN: DependencyN
    ): () -> Something = makeMyFunction(
            dependency1: Dependency1,
            ...
            dependencyN: DependencyN
    )
}
```

We can of course move implementation of `makeMyFunction()` to `MyFunctionModule` but then we introduce strong coupling of our DI agnostic code with Dagger. `makeMyFunction()` implementation can hold lets say an important business logic and it would be nice to not not mix this function with DI details. Adding `@Inject` annotation to it would still be "mixing with DI details" but to a much lesser extent than surrounding it with a module class. 

The added benefit of being able to add `@Inject` to top level functions is that instead of doing something like this.

```kotlin
class MyDependencyImpl @Inject constructor() : MyDependency {
    
}

@Module
interface MyDependencyModule {
    @Binds
    fun provideMyDependency(myDependency: MyDependencyImpl): MyDependency
}
```

We can now do it like this instead.

```kotlin
@Inject
fun makeMyDependency(): MyDependency = object : MyDependency {

}
```

So, there is no longer need to come up with a name for the implementation class and there is no longer need for a module with `@Binds` annotations. 

Basically the code bellow is no longer needed.

```kotlin
@Module
interface MyDependencyModule {
    @Binds
    fun provideMyDependency(myDependencyImpl: MyDependencyImpl): MyDependency
}
```

And for us it's a pretty common scenario when we have to do something like this. I think it would be nice to be able to not write this boilerplate anymore.

**Comment by JakeWharton** [comment]
Those seem far more like module-less providers than injection points. I can argue semantics, but I actually don't think it's that useful of a distinction in practice.

I'm extremely skeptical of how well this would work in a general sense. As soon as you get into things like scoping annotations you need the ability to associate a function with the corresponding component–especially since the function could live entirely separately from the component definition.

**Comment by tasomaniac** [comment]
Isn't that the same with constructor injection. When putting @Inject annotation to a class constructor you don't really associate it with a component either. Scopes are done via putting a scope annotation at class level.

**Comment by tbroyer** [comment]
What you're looking for is something akin to Guice's `@ProvidedBy`, but that annotation goes on the type being injected, which I don't think is possible here; or like CDI's [`@Produces`](https://docs.oracle.com/javaee/7/api/javax/enterprise/inject/Produces.html).
In any case, that would be something extra to JSR 330, not `@Inject`; so you cannot achieve "DI agnostic code", at least in the sense that it could work with another JSR 330 implementation than Dagger.

Fwiw, the way Dagger works, if it finds that it needs to inject a type `A`, it will look into the component's modules whether there's a specific binding for it, and will fall back to `A`'s `@Inject`-annotated constructor. It wouldn't be able to find your provider/factory method without somehow scanning the classpath (at runtime or compile-time, but still) for anything that would provide the expected type (*before* falling back to using its constructor). This is actually how CDI works *by design*, classpath scanning, autodiscovery, etc. the opposite of Dagger's mantra that things should be explicit (hey, it even requires you to annotate a zero-argument constructor with `@Inject`!)

You could probably have your own annotation and annotation processor (or Kotlin compiler plugin) to generate that module for you based on your annotated function (and possibly then a build plugin, like Hilt, for even fewer boilerplate). This is actually not much different from [AutoFactory](https://github.com/google/auto/tree/master/factory) or [AssistedInject](https://github.com/square/AssistedInject), except specific to Kotlin.
Or maybe when Dagger will have *native* support for factories (#1825) this would be somehow included?

**Comment by tasomaniac** [comment]
That definitely makes sense. Thanks @tbroyer for detailed explanation.

Btw, I feel like this can already be possible with Hilt by having `@file:InstallIn` and `@file:Module` annotations. I've not tried myself.

**Comment by guelo** [comment]
Injecting constructors with default parameters doesn't work.  The workarounds are to use the traditional Java telescoping constructor pattern and mark only one of them with `@Inject`. Or to put the default values into the Dagger graph.

An example that I keep running into is testing java.time stuff deterministically, I want to optionally pass in a Clock but I don't really want to get Dagger involved, I just want to write this. 
```kotlin
class UnderTest @Inject constructor (clock: Clock =  Clock.systemDefaultZone()) {
  ///
}
```
But that doesn't work, and `@JvmOverloads` doesn't help because it ends up tagging all the constructors with `@Inject`.

**Comment by ZacSweers** [comment]
@guelo that's more of a feature request. Java source code in general cannot leverage default parameter values. It's possible dagger could support this, but it's nontrivial as it would require both reflection at runtime and implicit handling similar to `@BindsOptionalOf`

**Comment by kaushalyap** [comment]
@JakeWharton , @ZacSweers Why not people develop compile time DI tool like Dagger 2 in Kotlin? (there only Koin and Kodein-DI which are service locators). Is it because project scope is big?
